builtin/game/item.lua

```diff
@@ -112,6 +112,61 @@ function core.facedir_to_dir(facedir)
 	return facedir_to_dir[facedir_to_dir_map[facedir % 32]]
 end
 
+local facedir_matrices = {}
+do
+	-- the following rotations are taken from rotateMeshBy6dFacedir in src/client/mesh.cpp
+	local m_dirs = {
+		[0] = matrix.identity,
+		matrix.rotation_around_x(math.pi / 2),
+		matrix.rotation_around_x(-math.pi / 2),
+		matrix.rotation_around_z(-math.pi / 2),
+		matrix.rotation_around_z(math.pi / 2),
+		matrix.rotation_around_z(-math.pi)
+	}
+	local m_rots = {
+		[0] = matrix.identity,
+		matrix.rotation_around_y(math.pi / 2),
+		matrix.rotation_around_y(math.pi),
+		matrix.rotation_around_y(math.pi * 3 / 2)
+	}
+	for facedir = 0, 23 do
+		local dir = math.floor(facedir / 4)
+		local rot = facedir % 4
+		facedir_matrices[facedir] = matrix.multiply(m_dirs[dir], m_rots[rot])
+	end
+end
+
+function core.facedir_to_matrix(facedir, nocopy)
+	return nocopy and facedir_matrices[facedir] or matrix.new(facedir_matrices[facedir])
+end
+
+local matrix_facedirs = {}
+
+-- returns a number asuming that the matrix has only -1, 0 and 1
+local function matrix_1p_1n_0_hash(m)
+	local r = 0
+	for i = 1, 9 do
+		r = r + (m[i] + 1) * 3^(i-1)
+		-- wrong values are possible, eg. m[i] could be 2. the following would fix that:
+		--~ if m[i] == 0 then
+			--~ r = r + 1 * 3^(i-1)
+		--~ elseif m[i] == 1 then
+			--~ r = r + 2 * 3^(i-1)
+		--~ elseif m[i] ~= -1 then
+			--~ return -1
+		--~ end
+	end
+	return r
+end
+
+for i = 0, 23 do
```

Comment:
Don't table indexes start from 1 in Lua?

Reply:
Yes, but formspec param2 values start at 0.

Reply:
I see. Thanks for clarifying.

Reply:
Perhaps it would be better if `facedir_matrices` was indexed from 1 to 24.
`matrix_facedirs`, however, has 0 to 23 as values, not keys.